### PR TITLE
[Python] Use thread-safe futures for concurrent operations

### PR DESCRIPTION
**Pull request by agners**
Instead of using quasi-global variables in the ChipStack singleton use device controller local futures to store results from callbacks. This has several advantages, namely:
- Avoid unnecessary shared state between device controllers
- Avoid unnecessary shared state between various operations within a device controller (those who don't share callbacks could be called from different threads now)
- Explicitly set Futures to None to detect spurious/unexpected callbacks
- Better code readability
- concurrent.futures are thread-safe
- Will make asyncio transition easier

This change shouldn't change the external API.

This allows us to get rid of the rather complicated `CallAsyncWithCompleteCallback`. While at it, also get rid of the largely unused `DCState` enum.

### CODE REVIEW

File: src/controller/python/chip/ChipDeviceCtrl.py [hunk 1]
```diff
@@ -1963,25 +2006,21 @@ def CommissionOnNetwork(self, nodeId: int, setupPinCode: int,
...
             filter = str(filter)
 
-        self._ChipStack.commissioningCompleteEvent.clear()
+        self._commissioning_complete_future = concurrent.futures.Future()
+        try:
+            self._enablePairingCompeleteCallback(True)
+            self._ChipStack.Call(
+                lambda: self._dmLib.pychip_DeviceController_OnNetworkCommission(
+                    self.devCtrl, self.pairingDelegate, nodeId, setupPinCode, int(filterType), str(filter).encode("utf-8") + b"\x00" if filter is not None else None, discoveryTimeoutMsec)
+            ).raise_on_error()
```
Comment (arkq) [inline src/controller/python/chip/ChipDeviceCtrl.py:2019]: Function was not raising before.

See other changed functions as well.

**Review by arkq — APPROVED** [review]

File: src/controller/python/chip/ChipDeviceCtrl.py [hunk 2]
```diff
@@ -1991,51 +2030,57 @@ def CommissionWithCode(self, setupPayload: str, nodeid: int, discoveryType: Disc
...
-        self.state = DCState.COMMISSIONING
+        self._commissioning_complete_future = concurrent.futures.Future()
 
-        self._ChipStack.commissioningCompleteEvent.clear()
+        try:
+            self._enablePairingCompeleteCallback(True)
+            self._ChipStack.Call(
+                lambda: self._dmLib.pychip_DeviceController_ConnectWithCode(
+                    self.devCtrl, setupPayload, nodeid, discoveryType.value)
+            ).raise_on_error()
```
[the remark this answers is not in the document]
Comment (agners) [reply src/controller/python/chip/ChipDeviceCtrl.py:2040]: Unfortunately it is a bit PITA and confusing to make it behave as before! I really appreciate a thorough review on this, as I rather prefer to keep API with this change. :pray: 

There are two operations which return `PyChipError`: Posting the command on the CHIP thread and calling it there and the actual return.

The posting onto the CHIP thread was handled in `CallAsyncWithCompleteCallback`, where this code was used:

```python
        res = self.PostTaskOnChipThread(callFunct).Wait()

        if not res.is_success:
            self.completeEvent.set()
            raise res.to_exception()
```

`Call()` just returns the `PyChipError`. So the  `.raise_on_error()` essentially makes this equivalent.

IMHO, we should convert the API to always raise exceptions on error, and simply return None. That is more pythonic. I intend to do this in a follow up. Maybe I should have done it first  :sweat_smile: :see_no_evil:

File: src/controller/python/chip/ChipDeviceCtrl.py [hunk 3]
```diff
@@ -529,26 +536,31 @@ def IsConnected(self):
...
-        if not self._ChipStack.commissioningCompleteEvent.isSet():
-            # Error 50 is a timeout
-            return PyChipError(CHIP_ERROR_TIMEOUT)
-        return self._ChipStack.commissioningEventRes
+        try:
+            self._enablePairingCompeleteCallback(True)
+            self._ChipStack.Call(
+                lambda: self._dmLib.pychip_DeviceController_ConnectBLE(
+                    self.devCtrl, discriminator, setupPinCode, nodeid)
+            ).raise_on_error()
```
[the remark this answers is not in the document]
Comment (agners) [reply src/controller/python/chip/ChipDeviceCtrl.py:546]: Yeah, that seems wrong type annotation :see_no_evil: 

That said, I like to keep the behavior right now. But it seems I made an error here: The `_commissioning_complete_future` does not raise exceptions *currenty*. So I need to add a `.raise_on_error()` for the result in this case.

File: src/controller/python/chip/ChipDeviceCtrl.py [hunk 4]
```diff
@@ -582,32 +594,62 @@ def CloseSession(self, nodeid):
...
+            self._ChipStack.Call(
+                lambda: self._dmLib.pychip_DeviceController_EstablishPASESessionIP(
+                    self.devCtrl, ipaddr.encode("utf-8"), setupPinCode, nodeid, port)
+            ).raise_on_error()
+
+            # TODO: This is a bit funky, but what the API returned with the previous
+            # implementation. We should revisit this.
+            err = self._pase_establishment_complete_future.result()
+            if not err.is_success:
+                return err.to_exception()
```
[the remark this answers is not in the document]
Comment (agners) [reply src/controller/python/chip/ChipDeviceCtrl.py:629]: `raise_on_error()` raises an exception, my code returns an exception.

But yeah, this is super ugly, but that is what it was before :see_no_evil: :cry:  So I did just the same :see_no_evil:

[the remark this answers is not in the document]
Comment (arkq) [reply src/controller/python/chip/ChipDeviceCtrl.py:633]: Ooo yes, my mistake!

File: src/controller/python/chip/ChipDeviceCtrl.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (arkq) [reply src/controller/python/chip/ChipDeviceCtrl.py:2044]: > IMHO, we should convert the API to always raise exceptions on error, and simply return None. That is more pythonic.

100% agree on that!

File: src/controller/python/chip/ChipDeviceCtrl.py (hunk 1 above)
Comment (agners) [reply src/controller/python/chip/ChipDeviceCtrl.py:2023]: Similar to the first instance, it was through `CallAsyncWithCompleteCallback`. Marking as resolved as well.

**Review by cecille — APPROVED** [review]

**Pull request merged**